man/pstops-clip.1.in
.\" Copyright (C) 2001-2012 Peter Selinger.
.\" This file is part of the upprint package. It is free software and
.\" is distributed under the terms of the GNU general public license.
.\" See the file COPYING for details.
.\" 
.TH PSTOPS-CLIP 1 "@DATE@" "Version @VERSION@" ""
.SH NAME
pstops-clip \- shuffle pages in a PostScript file
.SH SYNOPSIS
.B pstops-clip
[
.I options
] 
.I pagespecs
[
.I infile
[
.I outfile
] ]
.SH DESCRIPTION
.I Pstops-clip
rearranges pages from a PostScript document, creating a new PostScript file.
The input PostScript file should follow the Adobe Document Structuring
Conventions.
.I Pstops-clip
can be used to perform a large number of arbitrary re-arrangements of
Documents, including arranging for printing 2-up, 4-up, booklets, reversing,
selecting front or back sides of documents, scaling, etc. It is derived
from Angus J. C. Duggan\'s
.I pstops
program, with the added ability of specifying a clippath for each page. 
.PP
.I pagespecs
follow the syntax:
.RS 5
.TP 10
.I pagespecs
= [\fImodulo\fP:]\fIspec\fP[+\fIspec\fP|,\fIspec\fP]...
.TP
.I spec
= [-]\fIpageno\fP[L|R|U][@\fIscale\fP][(\fIxoff\fP,\fIyoff\fP)][{\fIx0\fP,\fIy0\fP,\fIx1\fP,\fIy1\fP}]
.PD
.RE
.sp
.TP 10
.I modulo
is the number of pages in each block. The value of
.I modulo
should be greater than 0; the default value is 1.
.TP
.I spec
is the page specification for one page in each block. 
If page \fIspec\fPs are separated by
.B \+
the pages will be merged into one page; if they are separated by
.B ,
they will be on separate pages.
.TP
.I pageno
This value specifies the relative page number within a block, and
should be between 0 (for the first page in the block) and \fImodulo\fP-1
(for the last page in each block) inclusive.
If the optional minus sign is specified, the page is relative to the end of
the document, instead of the start.
If there is only one page specification, with
.I pageno
zero, the \fIpageno\fP may be omitted.
.TP
.I scale
This optional parameter scales the page by the fraction specified.
.TP
\fIL\fP, \fIR\fP, \fIU\fP
These optional parameters rotate the page left,
right, or upside-down.
.TP
\fIxoff\fP, \fIyoff\fP
These optional dimensions shift the page by the specified amount.
They are in PostScript's points, but may be followed by the units
.B "cm"
or
.B "in"
to convert to centimetres or inches, or the flag
.B "w"
or
.B "h"
to specify as a multiple of the width or height.
.TP
\fIx0\fP, \fIy0\fP, \fIx1\fP, \fIy1\fP
These optional dimensions are used to specify a custom clipping path for
the page. 
.PD
.PP
The shift, rotation, and scaling are performed
in that order regardless of which order
they appear on the command line.
.SH OPTIONS
The following options are supported:
.TP 11
.B --help
Print usage information and exit.
.TP
.B --version, -v
Print version information and exit.
.TP
.B --license
Print license information and exit.
.TP
\fB\-w\fP\fIwidth\fP
Set the width which is used by the
.B "w"
dimension specifier. These dimensions are also used (after scaling) to set the
clipping path for each page.
.TP
\fB\-h\fP\fIheight\fP
Set the height which is used by the
.B "h"
dimension specifier. These dimensions are also used (after scaling) to set the
clipping path for each page.
.TP
\fB\-p\fP\fIpaper\fP
An alternative to \fB\-w\fP and \fB\-h\fP. Set the paper size to
.B a3, a4, a5, b5, letter, legal, tabloid, statement, executive, folio, quarto
or
.B 10x14.
The default paper size is
.B @DEFAULT_FORMAT@.
.TP
.B \-b
Prevent any
.B bind
operators in the PostScript prolog from binding. This may be needed in cases
where complex multi-page re-arrangements are being done.
.TP
\fB\-d\fP[\fIlinewidth\fP]
Draw a line around the border of each page, of the specified width.
If the \fIlinewidth\fP parameter is omitted, a default linewidth of 1 point is
assumed. The linewidth is relative to the original page dimensions,
\fIi.e.\fP it is scaled up or down with the rest of the page.
.TP
.B \-q
Pstops-clip normally prints the page numbers of the pages re-arranged; the
.I \-q
option suppresses this.
.PD
.SH EXAMPLES
This section contains some sample re-arrangements. To put two pages on one
sheet (of A4 paper), the pagespec to use is:
.sp
.ce
2:0L@.7(21cm,0)+1L@.7(21cm,14.85cm)
.sp
To select all of the odd pages in reverse order, use:
.sp
.ce
2:-0
.sp
To re-arrange pages for printing 2-up booklets, use
.sp
.ce
4:-3L@.7(21cm,0)+0L@.7(21cm,14.85cm)
.sp
for the front sides, and
.sp
.ce
4:1L@.7(21cm,0)+-2L@.7(21cm,14.85cm)
.sp
for the reverse sides (or join them with a comma for duplex printing).
.SH AUTHORS
Copyright (C) Angus J. C. Duggan 1991-1995
.br
Copyright (C) Peter Selinger 2006-2012
.SH "SEE ALSO"
\fBpsdim\fP(1), \fBlprwrap\fP(1)
.SH TRADEMARKS
.B PostScript
is a trademark of Adobe Systems Incorporated.
.SH BUGS
.I Pstops-clip
does not accept all DSC comments.


NEWS
March 27, 2012: upprint, release 1.7

 Portability improvements. Lprwrap: pass unrecognized -o options to
 the underlying lpr, so that applications expecting a CUPS lpr will
 still work. Changed the way lprwrap calls the "real" lpr.

July 23, 2011: upprint, release 1.6

 Improved portability. Lprwrap: Compatibility with newer versions of
 CUPS has been improved. The -ocups option to lprwrap now replaces the
 old lprwrap-cups script. Allow multiple -ofudge options. New --lpr
 option can be used to specify the backend "lpr" executable. The
 TEMPDIR environment variable is now respected. Support for pdftops
 added.  Added a size limit on text files, and a --force option to
 override it. Psdim: when compiled with metric defaults, use metric
 units in the output format string. New options --[lrtb]adjust for
 adjusting the bounding box of input document. Compensate for certain
 PostScript files generated by Acrobat Reader that resulted in shifted
 output.

February 15, 2007: upprint, release 1.5

 The new upprint package combines the features of the old lpr-wrapper
 and psdim packages. It contains lprwrap, psdim, as well as the new
 program pstops-clip. Page clipping was improved, so that documents
 with opaque backgrounds can now be printed in n-up mode. A 'fudge'
 feature was added to allow more flexible margin detection.  The
 lpr-wrapper program was renamed lprwrap.

January 22, 2007: lpr-wrapper, release 0.7

 Better CUPS support was added. PPD printer description files are no
 longer required. Printing of Unicode text files is now supported via
 iconv.

March 30, 2005: psdim, release 1.4

 This is a maintenance release. The sources have been updated to
 compile with recent versions of gcc.

January 15, 2004: lpr-wrapper, release 0.6

 We now treat any unknown file type consisting of printable characters
 as text. This allows printing text-based files such as GPG keys,
 which are not identified as "text" by the "file" program. We also
 call the backend "lpr" with the "-l" option to prevent filtering in
 CUPS. 

December 6, 2003: psdim, release 1.3

 This release does not differ from previous releases in functionality.
 The sources have been restructured to take advantage of standard
 automake/autoconf configuration. Configuration options --enable-a4
 and --enable-metric have been added. A spurious source of "Postscript
 errors" has been eliminated.


pstops-clip/pstops-clip.c
/* pstops.c
 * Copyright (C) Angus J. C. Duggan 1991-1995
 * Copyright (C) Peter Selinger 2006-2012
 * See file LICENSE for details.
 *
 * rearrange pages in conforming PS file for printing in signatures
 *
 * Usage:
 *       pstops [-q] [-b] [-d] [-w<dim>] [-h<dim>] [-ppaper] <pagespecs> [infile [outfile]]
 */

#ifdef HAVE_CONFIG_H
 #include "config.h"
#endif

#include <string.h>
#include <stdio.h>
#include <libgen.h>

#include "psutil.h"
#include "psspec.h"
#include "pserror.h"

char *program ;
int pages ;
int verbose ;
FILE *infile ;
FILE *outfile ;
char pagelabel[BUFSIZ] ;
int pageno ;

static void license(FILE *f) {
   fprintf(f, 
"%s "VERSION".\n"
"Copyright (C) 1991-1995 Angus J. C. Duggan.\n"
"Copyright (C) 2006-2012 Peter Selinger.\n"
"\n"
"This program may be copied and used for any purpose (including\n"
"distribution as part of a for-profit product), provided:\n"
"\n"
"1) The original attribution of the programs is clearly displayed in\n"
"   the product and/or documentation, even if the programs are modified\n"
"   and/or renamed as part of the product.\n"
"\n"
"2) The original source code of the programs is provided free of charge\n"
"   (except for reasonable distribution costs). For a definition of\n"
"   reasonable distribution costs, see the Gnu General Public License\n"
"   or Larry Wall's Artistic License (provided with the Perl 4 kit). The\n"
"   GPL and Artistic License in NO WAY affect this license; they are merely\n"
"   used as examples of the spirit in which it is intended.\n"
"\n"
"3) These programs are provided \"as-is\". No warranty or guarantee of\n"
"   their fitness for any particular task is provided. Use of these\n"
"   programs is completely at your own risk.\n"
	   , program);
}

static void usage(FILE *f) {
   fprintf(f, 
"pstops-clip "VERSION". Rearranges pages from a PostScript document.\n"
"\n"
"Usage: %s [options] <pagespecs> [infile [outfile]]\n"
"Options:\n"
" --help               - print this help message and exit\n"
" --version, -v        - print version info and exit\n"
" --license            - print license info and exit\n"
" -q                   - don't print page numbers to stderr\n"
" -b                   - prevent PostScript 'bind' operators from binding\n"
" -w<width>            - set page width\n"
" -h<height>           - set page height\n"
" -p<paper>            - set paper size\n"
" -d[linewidth]        - draw a box around each page\n"
"\n"
"Paper sizes:\n"
" a3, a4, a5, b5, letter, legal, tabloid, statement, executive, folio,\n"
" quarto, 10x14, _glibc\n"
"\n"
"Page specification:\n"
" <pagespecs>  = [modulo:]<spec>[,spec|+spec]...\n"
" <spec>       = [-]pageno[@scale][L|R|U][(xoff,yoff)][{x0,y0,x1,y1}]\n"
" modulo:      the number of pages in each block (1 or more; default 1)\n"
" spec:        the page specification for one page in each block\n"
" pageno:      between 0 and modulo-1\n"
" scale:       scaling factor\n"
" L,R,U:       rotate left, right, upside-down\n"
" xoff,yoff:   page origin in PostScript points, or units of cm, in, w, h\n"
" x0,y1,x1,y1: page clip path in PostScript points, or units of cm, in, w, h\n"
"\n"
	   , program);
}

void version(FILE *f)
{
   fprintf(f, 
"%s "VERSION".\n"
"Copyright (C) Angus J. C. Duggan, 1991-1995. See --license for details.\n"
"Copyright (C) 2006-2012 Peter Selinger.\n"
	   , program);
}

/* this can be used as an error handling callback function; does not return */
void shortusage(void)
{
   fprintf(stderr, "Illegal command line. Try --help for more info.\n");
   fflush(stderr);
   exit(1);
}

static int modulo = 1;
static int pagesperspec = 1;

static PageSpec *parsespecs(char *str)
{
   PageSpec *head, *tail;
   int other = 0;
   int num = -1;

   head = tail = newspec();
   while (*str) {
      if (isdigit(*str)) {
	 num = parseint(&str, shortusage);
      } else {
	 switch (*str++) {
	 case ':':
	    if (other || head != tail || num < 1) shortusage();
	    modulo = num;
	    num = -1;
	    break;
	 case '-':
	    tail->reversed = !tail->reversed;
	    break;
	 case '@':
	    if (num < 0) shortusage();
	    tail->scale *= parsedouble(&str, shortusage);
	    tail->flags |= SCALE;
	    break;
	 case 'l': case 'L':
	    tail->rotate += 90;
	    tail->flags |= ROTATE;
	    break;
	 case 'r': case 'R':
	    tail->rotate -= 90;
	    tail->flags |= ROTATE;
	    break;
	 case 'u': case 'U':
	    tail->rotate += 180;
	    tail->flags |= ROTATE;
	    break;
	 case '(':
	    tail->xoff += parsedimen(&str, shortusage);
	    if (*str++ != ',') shortusage();
	    tail->yoff += parsedimen(&str, shortusage);
	    if (*str++ != ')') shortusage();
	    tail->flags |= OFFSET;
	    break;
	 case '{':
	    tail->x0 = parsedimen(&str, shortusage);
	    if (*str++ != ',') shortusage();
	    tail->y0 = parsedimen(&str, shortusage);
	    if (*str++ != ',') shortusage();
	    tail->x1 = parsedimen(&str, shortusage);
	    if (*str++ != ',') shortusage();
	    tail->y1 = parsedimen(&str, shortusage);
	    if (*str++ != '}') shortusage();
	    tail->flags |= CLIP;
	    break;
	 case '+':
	    tail->flags |= ADD_NEXT;
	 case ',':
	    if (num < 0 || num >= modulo) shortusage();
	    if ((tail->flags & ADD_NEXT) == 0)
	       pagesperspec++;
	    tail->pageno = num;
	    tail->next = newspec();
	    tail = tail->next;
	    num = -1;
	    break;
	 default:
	    shortusage();
	 }
	 other = 1;
      }
   }
   if (num >= modulo)
      shortusage();
   else if (num >= 0)
      tail->pageno = num;
   return (head);
}

int main(int argc, char *argv[])
{
   PageSpec *specs = NULL;
   int nobinding = 0;
   double draw = 0;
   Paper *paper;

#ifdef PAPER
   if ( (paper = findpaper(PAPER)) != (Paper *)0 ) {
      width = (double)PaperWidth(paper);
      height = (double)PaperHeight(paper);
   }
#endif

   infile = stdin;
   outfile = stdout;
   verbose = 1;
   for (program = basename(*argv++); --argc; argv++) {
      if (strcmp(argv[0], "--help") == 0) {
	 usage(stdout);
	 exit(1);
      } else if (strcmp(argv[0], "--version") == 0) {
	 version(stdout);
	 exit(1);
      } else if (strcmp(argv[0], "--license") == 0) {
	 license(stdout);
	 exit(1);
      } else if (argv[0][0] == '-') {
	 switch (argv[0][1]) {
	 case 'q':	/* quiet */
	    verbose = 0;
	    break;
	 case 'd':	/* draw borders */
	    if (argv[0][2])
	       draw = singledimen(*argv+2, shortusage, shortusage);
	    else
	       draw = 1;
	    break;
	 case 'b':	/* no bind operator */
	    nobinding = 1;
	    break;
	 case 'w':	/* page width */
	    width = singledimen(*argv+2, shortusage, shortusage);
	    break;
	 case 'h':	/* page height */
	    height = singledimen(*argv+2, shortusage, shortusage);
	    break;
	 case 'p':	/* paper type */
	    if ( (paper = findpaper(*argv+2)) != (Paper *)0 ) {
	       width = (double)PaperWidth(paper);
	       height = (double)PaperHeight(paper);
	    } else
	      message(FATAL, "paper size '%s' not recognised\n", *argv+2);
	    break;
	 case 'v':	/* version */
	    version(stdout);
	    exit(1);
	 default:
	    if (specs == NULL)
	       specs = parsespecs(*argv);
	    else
	       shortusage();
	 }
      } else if (specs == NULL)
	 specs = parsespecs(*argv);
      else if (infile == stdin) {
	 if ((infile = fopen(*argv, "r")) == NULL)
	    message(FATAL, "can't open input file %s\n", *argv);
      } else if (outfile == stdout) {
	 if ((outfile = fopen(*argv, "w")) == NULL)
	    message(FATAL, "can't open output file %s\n", *argv);
      } else shortusage();
   }
   if (specs == NULL)
      shortusage();
#if defined(MSDOS) || defined(WINNT)
   if ( infile == stdin ) {
      int fd = fileno(stdin) ;
      if ( setmode(fd, O_BINARY) < 0 )
         message(FATAL, "can't open input file %s\n", argv[4]);
    }
   if ( outfile == stdout ) {
      int fd = fileno(stdout) ;
      if ( setmode(fd, O_BINARY) < 0 )
         message(FATAL, "can't reset stdout to binary mode\n");
    }
#endif
   if ((infile=seekable(infile))==NULL)
      message(FATAL, "can't seek input\n");

   pstops(modulo, pagesperspec, nobinding, specs, draw);

   exit(0);
}
